src/components/TradeWidget/OrderValidity.tsx

```diff
@@ -247,56 +415,124 @@ interface Props {
 const OrderValidity: React.FC<Props> = ({
   isDisabled,
   isUnlimited,
-  setAsap,
-  setUnlimited,
   isAsap,
   tabIndex,
   validFromInputId,
   validUntilInputId,
 }) => {
   const [showOrderConfig, setShowOrderConfig] = useSafeState(false)
-  const { setValue, errors, register, getValues, watch } = useFormContext<TradeFormData>()
-  const { validFrom, validUntil } = getValues()
+
+  const formMethods = useFormContext<TradeFormData>()
+  const { setValue, errors, register, watch, trigger } = formMethods
+  const { validFrom: validFromTimeMs, validUntil: validUntilTimeMs } = watch([validFromInputId, validUntilInputId])
+
+  // const validFromDefault = validFromBatchId ? Date.parse(batchIdToDate(+validFromBatchId).toString()) : null
+  const [validFromButton, setValidFromButton] = useSafeState<number | null>(
+    isAsap ? null : validFromTimeMs ? Infinity : VALID_FROM_DEFAULT,
+  )
+  const [validFromCustomTime, setValidFromCustomTime] = useSafeState<number | null>(
+    (validFromTimeMs && +validFromTimeMs) || null,
+  )
+  const [validUntilButton, setValidUntilButton] = useSafeState<number | null>(
+    isUnlimited ? null : validUntilTimeMs ? Infinity : VALID_UNTIL_DEFAULT,
+  )
+  const [validUntilCustomTime, setValidUntilCustomTime] = useSafeState<number | null>(
+    (validUntilTimeMs && +validUntilTimeMs) || null,
+  )
+
+  const [validFromCustomBatchId, setValidFromCustomBatchId] = useSafeState<string | null>(
+    validFromCustomTime ? dateToBatchId(validFromCustomTime).toString() : null,
+  )
+  const [validUntilCustomBatchId, setValidUntilCustomBatchId] = useSafeState<string | null>(
+    validUntilCustomTime ? dateToBatchId(validUntilCustomTime).toString() : null,
+  )
 
   const validFromError = errors[validFromInputId]
   const validUntilError = errors[validUntilInputId]
-  const validFromInputValue = watch(validFromInputId)
-  const validUntilInputValue = watch(validUntilInputId)
-  const overMax = ZERO
-  const validFromClassName = validFromError ? 'error' : overMax.gt(ZERO) ? 'warning' : ''
-  const validUntilClassName = validUntilError ? 'error' : overMax.gt(ZERO) ? 'warning' : ''
 
-  const handleShowConfig = useCallback((): void => {
+  const handleRelativeTimeSelect = useCallback(
```

Comment:
Can you add comments to these `handle*` functions and a bigger comment above the component with what states depends on what.
Like a reactive chain
```
validFrom -> validUntil -> customValidUntilDate
               \
                -> customValidFromDate
```

to help wrap one's head around interdependencies

Reply:
will do